File: backend/internal/service/tests_17.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// LabTestRecord17Service encapsulates business rules for LabTestRecord17.
type LabTestRecord17Service struct {
	repo *repository.LabTestRecord17Repository
}

// NewLabTestRecord17Service wires the service.
func NewLabTestRecord17Service(repo *repository.LabTestRecord17Repository) *LabTestRecord17Service {
	return &LabTestRecord17Service{repo: repo}
}

// Create validates and persists a new LabTestRecord17.
func (s *LabTestRecord17Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.LabTestRecord17, error) {
	now := time.Now().UTC()
	rec := &models.LabTestRecord17{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a LabTestRecord17 by id.
func (s *LabTestRecord17Service) Get(ctx context.Context, id string) (*models.LabTestRecord17, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped LabTestRecord17 pages.
func (s *LabTestRecord17Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.LabTestRecord17, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing LabTestRecord17.
func (s *LabTestRecord17Service) Update(ctx context.Context, actor string, rec *models.LabTestRecord17) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a LabTestRecord17.
func (s *LabTestRecord17Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds LabTestRecord17 by external code prefix.
func (s *LabTestRecord17Service) Search(ctx context.Context, facilityID, prefix string) ([]models.LabTestRecord17, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *LabTestRecord17Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/tests17.ts
/** Domain types for tests / LabTestRecord17 */
export type LabTestRecord17Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface LabTestRecord17 {
  id: string;
  externalCode: string;
  displayName: string;
  status: LabTestRecord17Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface LabTestRecord17CreateRequest {
  externalCode: string;
  displayName: string;
  status: LabTestRecord17Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface LabTestRecord17ListResponse {
  items: LabTestRecord17[];
  total?: number;
}

export interface LabTestRecord17StatsResponse {
  activeCount: number;
}

export function isLabTestRecord17Active(rec: LabTestRecord17): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatLabTestRecord17Label(rec: LabTestRecord17): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareLabTestRecord17ByPriority(a: LabTestRecord17, b: LabTestRecord17): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
